onnxruntime/core/graph/contrib_ops/contrib_defs.cc

```diff
@@ -172,6 +172,97 @@ activation.)DOC")
+      .Attr(
+          "M",
+          "Minimum n-gram length",
+          AttributeProto::INT)
+      .Attr(
+          "S",
+          "Maximum number of items(integers/strings) to be skipped when constructing an n-gram from X.",
+          AttributeProto::INT)
+      .Attr(
+          "pool_strings",
+          "List of strings n-grams learned from the training set. Either this or pool_int64s attributes must be present but not both."
+          "It's an 1-D tensor starting with the collections of all 1-grams and ending with the collections of n-grams."
+          "The i-th element in pool stores the n-gram that should be mapped to index ngram_indexes[i] in the output vector.",
+          AttributeProto::STRINGS,
+          OPTIONAL)
+      .Attr(
+          "pool_int64s",
+          "List of int64 n-grams learned from the training set. Either this or pool_strings attributes must be present but not both."
+          "It's an 1-D tensor starting with the collections of all 1-grams and ending with the collections of n-grams."
+          "The i-th element in pool stores the n-gram that should be mapped to index ngram_indexes[i] in the output vector.",
+          AttributeProto::INTS,
+          OPTIONAL)
+      .Attr(
+          "ngram_counts",
+          "The starting indexes of 1-grams, 2-grams, and so on in pool."
+          "It is useful when determining the boundary between two consecutive collections of n-grams."
+          "For example, if ngram_counts is [0, 17, 36], the first index (zero-based) of 1-gram/2-gram/3-gram"
+          "in pool are 0/17/36",
+          AttributeProto::INTS)
+      .Attr(
+          "ngram_indexes",
+          "list of int64s (type: AttributeProto::INTS). This list is parallel to the specified 'pool_*' attribute."
+          "The i-th element in ngram_indexes indicate the coordinate of the i-th n-gram in the output tensor.",
+          AttributeProto::INTS)
+      .Attr(
+          "weights",
+          "list of floats. This attribute stores the weight of each n-gram in pool. The i-th element in weights"
+          "is the weight of the i-th n-gram in pool. Its length equals to the size of ngram_indexes."
+          "By default, weights is an all-one tensor.This attribute is used when mode is \"IDF\" or \"TFIDF\""
+          "to scale the associated word counts.",
+          AttributeProto::FLOATS,
+          OPTIONAL)
+      .Attr(
+          "mode",
+          "The weighting criteria. It can be one of \"TF\" (term frequency),"
+          "\"IDF\" (inverse document frequency), and \"TFIDF\" (the combination of TF and IDF)",
+          AttributeProto::STRING)
+      .TypeAndShapeInferenceFunction([](ONNX_NAMESPACE::InferenceContext& ctx) {
+        auto output_elem_type = ctx.getOutputType(0)->mutable_tensor_type();
+        output_elem_type->set_elem_type(ONNX_NAMESPACE::TensorProto::FLOAT);
+      })
+      .SetDoc(R"DOC(
+This transform extracts n-grams from the input sequence and save them as a vector.
```

Comment:
```suggestion
 This transform extracts n-grams from the input sequence and save them as a vector. Input can
 be either a 1-D or 2-D tensor. For 1-D input, output is the n-gram representation of that input. 
 For 2-D input, the output is also a  2-D tensor whose i-th row is the n-gram representation f the i-th input row.
```

Reply:
Thanks! Will change accordingly.

---
In reply to: [243379535](https://github.com/Microsoft/onnxruntime/pull/180#discussion_r243379535) [](ancestors = 243379535)